### PR TITLE
Added a validator for Read-Only files.

### PR DESCRIPTION
**Pull request by FelixGV**
Given a data directory, the ```ReadOnlyFileValidator``` class will check a few things:
- Index files are well formed:
  - Well sorted
  - no negative offset
- Data files are well formed:
  - Number of records for a given key hash is > 0
  - Key and value sizes are > 0
- Index and data files match together:
  - No gaps in data files, when traversing them with the offsets from index files
  - Raw keys in data files match the hashes found in their index files

It then prints a summary of the amount and kinds of problems found.

I am posting this here in case anyone is curious, but I might modify it further before being ready to merge it in, so please don't merge it in yet yourselves. Thanks.

### CODE REVIEW

**Comment by arunthirupathi** [comment]
Do you think a check to read .metadata file and make sure the checksum matches between the data and index file is worth adding ?

That will help in identifying the disk corruption.

The way I see it is , these checks if errors out we have errors. But absence of errors does not mean they are good.

File: src/java/voldemort/store/readonly/ReadOnlyFileValidator.java [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+        for (File file: files) {
+            if (file.length() == 0) {
+                if (verbose) {
+                    System.out.println("Skipping empty file: " + file.getName());
+                }
+            } else if (file.getName().endsWith(ReadOnlyUtils.INDEX_FILE_EXTENSION)) {
+                problemsDetectedInIndexFiles += validateIndexFile(file, verbose);
+                problemsDetectedInIndexAndDataFilesMatch += validateIndexAndDataFileMatch(file, verbose);
+            } else if (file.getName().endsWith(ReadOnlyUtils.DATA_FILE_EXTENSION)) {
+                problemsDetectedInDataFiles += validateDataFile(file, verbose);
```
Comment (arunthirupathi) [inline src/java/voldemort/store/readonly/ReadOnlyFileValidator.java]: Should you check for existence of .index file here ?

File: src/java/voldemort/store/readonly/ReadOnlyFileValidator.java [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+                                                       String description,
+                                                       int keysRead) {
+        int cmp = ByteUtils.compare(previousByteArray, currentByteArray);
+
+        if (cmp < 0) {
+            // This is the expected case. The index file appears to be well-formed so far.
+            return 0;
+        } else if (cmp == 0) {
+            // The same key hash or offset appeared twice in a row. Unexpected!
+            System.err.println("We found two identical consecutive key " + description +
```
Comment (arunthirupathi) [inline src/java/voldemort/store/readonly/ReadOnlyFileValidator.java]: Should we print them only under verbose conditions ?

Comment (FelixGV) [inline src/java/voldemort/store/readonly/ReadOnlyFileValidator.java]: Yes, good point.

File: src/java/voldemort/store/readonly/ReadOnlyFileValidator.java [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+            valueSize = ByteBuffer.wrap(currentValueSize).getInt();
+
+            if (verbose) {
+                System.out.println("Key #" + keysRead +
+                                   ", Key Size: " + keySize +
+                                   ", Value Size: " + valueSize);
+            }
+
+            if (keySize == 0) {
+                System.out.println("Key size 0!");
```
Comment (arunthirupathi) [inline src/java/voldemort/store/readonly/ReadOnlyFileValidator.java]: Same verbose here, also this uses System.out other uses System.err

Comment (FelixGV) [inline src/java/voldemort/store/readonly/ReadOnlyFileValidator.java]: Good catch, I'll refactor all of that.

**Comment by FelixGV** [comment]
Thanks for the review @arunthirupathi. I amended my previous commit.

I added a bash script to easily run the ```ReadOnlyFileValidator```. I also added disclaimers about what this script does not do.

I'm good to go as far as merging this in.

**Comment by arunthirupathi** [comment]
Looks good to me

**Pull request merged**